torch_brain/models/poyo.py

```diff
@@ -0,0 +1,354 @@
+        num_latents=64,
+        depth=2,
+        cross_heads=1,
+        self_heads=8,
+        ffn_dropout=0.2,
+        lin_dropout=0.4,
+        atn_dropout=0.0,
+        emb_init_scale=0.02,
+        t_min=1e-4,
+        t_max=4.0,
+    ):
+        super().__init__()
+
+        # embeddings
+        self.unit_emb = InfiniteVocabEmbedding(dim, init_scale=emb_init_scale)
+        self.session_emb = InfiniteVocabEmbedding(dim, init_scale=emb_init_scale)
+        self.token_type_emb = Embedding(4, dim, init_scale=emb_init_scale)
+        self.latent_emb = Embedding(num_latents, dim, init_scale=emb_init_scale)
+        self.rotary_emb = RotaryEmbedding(dim_head, t_min, t_max)
+
+        self.dropout = nn.Dropout(p=lin_dropout)
+
+        # encoder layer
+        self.enc_atn = RotaryCrossAttention(
+            dim=dim,
+            heads=cross_heads,
+            dropout=atn_dropout,
+            dim_head=dim_head,
+            rotate_value=True,
+        )
+        self.enc_ffn = nn.Sequential(
+            nn.LayerNorm(dim), FeedForward(dim=dim, dropout=ffn_dropout)
+        )
+
+        # process layers
+        self.proc_layers = nn.ModuleList([])
+        for i in range(depth):
+            self.proc_layers.append(
+                nn.Sequential(
+                    RotarySelfAttention(
+                        dim=dim,
+                        heads=self_heads,
+                        dropout=atn_dropout,
+                        dim_head=dim_head,
+                        rotate_value=True,
+                    ),
+                    nn.Sequential(
+                        nn.LayerNorm(dim),
+                        FeedForward(dim=dim, dropout=ffn_dropout),
+                    ),
+                )
+            )
+
+        # decoder layer
+        self.dec_atn = RotaryCrossAttention(
+            dim=dim,
+            heads=cross_heads,
+            dropout=atn_dropout,
+            dim_head=dim_head,
+            rotate_value=False,
+        )
+        self.dec_ffn = nn.Sequential(
+            nn.LayerNorm(dim), FeedForward(dim=dim, dropout=ffn_dropout)
+        )
+
+        # Output projections + loss
+        self.readout = nn.Linear(dim, dim_out)
+
+        self.dim = dim
+
+    def forward(
+        self,
+        *,
+        # input sequence
+        input_unit_index: TensorType["batch", "n_in", int],
+        input_timestamps: TensorType["batch", "n_in", float],
+        input_token_type: TensorType["batch", "n_in", int],
+        input_mask: Optional[TensorType["batch", "n_in", bool]] = None,
+        # latent sequence
+        latent_index: TensorType["batch", "n_latent", int],
+        latent_timestamps: TensorType["batch", "n_latent", float],
+        # output sequence
+        output_session_index: TensorType["batch", "n_out", int],
+        output_timestamps: TensorType["batch", "n_out", float],
+    ) -> TensorType["batch", "n_out", "dim_out", float]:
+        """Forward pass of the POYO+ model.
```

Comment:
To clarify, I am not referring to chain vs. pad. What I am suggesting is replicating the `unpack_output=True` behavior, so that instead of returning `B x T_out x D_out`, we return `List[*, D_out]` of length `B`. The logic for this is super simple. The typical user is not aware that masking is being used! — it's happening in the tokenizer/collate. so they wouldn't know that the output they are getting will need to be "unpadded", in which case they would have to know how to access to the `output_mask` inside of the batch dict. This would of course happen rarely but it is likely incase some dataset has missing values etc...

My suggestion is to include a single line of code:
```
if unpack_output:
     return [output[b][output_mask[b]] for b in range(output.size(0))]
```

Reply:
Fair point. Done.